torch/csrc/jit/ir.h

```diff
@@ -72,28 +72,107 @@ struct SourceLocation {
 // the list types are intentionally simple, but we type-def
 // them here so if we need to change them, refactoring will be easier
 using node_list = std::vector<Node*>;
-using param_list = node_list;
+using value_list = std::vector<Value*>;
 using use_list = std::vector<Use>;
 using pyobj_list = std::vector<THPObjectPtr>;
 template<typename T>
 using ArrayRef = at::ArrayRef<T>;
 using NodeKind = Symbol;
 
-inline TypePtr getInitialType(NodeKind kind) {
-  switch(kind) {
-    case kPythonOp:
-    case kCppOp:
-    case kEval:
-    case kFusionGroup:
-      return multiType();
-    default:
-      return nullptr;
+struct Value {
+  TH_DISALLOW_COPY_AND_ASSIGN(Value);
+  Value(Node * node_, size_t offset_);
+private:
+  friend struct Node;
+  Node * node_;
+  size_t offset_;
+  size_t unique_ = 0;          // unique id
+  size_t stage_ = 0;           // 0-forward, 1-backward, 2-double-backward,...
```

Comment:
(This comment requests no changes for the current PR). I do idly wonder if you might be able to kill this by introducing a synthetic node whose outputs are all of the actual input parameters; a kind of "load input" instruction (ala Swift IR). It may be hard to see if this is actually an improvement unless we actually do the change. Something to keep an eye out for if maintaining stage in both Value and Node becomes painful.

Reply:
Yeah, if we separate inputs per stage into different "Param" instructions then we can drop stage from inputs. It would clean up the staging, at the cost of making things like ONNX which disregard the stage more confusing since outputs/inputs to the graph would be a list of lists.  So not for now, but maybe later.